### PR TITLE
Adding http io trasport implementation

### PR DESCRIPTION
**Pull request by Yasara123**
Adding the http source and sink implementation using http carbon transport and OSGi test component.

### CODE REVIEW

File: component/pom.xml [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            <groupId>org.wso2.siddhi</groupId>
+            <artifactId>siddhi-annotations</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.siddhi</groupId>
+            <artifactId>siddhi-extension-xml-input-mapper</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.siddhi</groupId>
+            <artifactId>siddhi-service</artifactId>
```
Comment (suhothayan) [inline component/pom.xml]: Why we need this ?

File: component/pom.xml [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        <dependency>
+            <groupId>org.wso2.siddhi</groupId>
+            <artifactId>siddhi-extension-json-input-mapper</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon</groupId>
+            <artifactId>org.wso2.carbon.logging</artifactId>
+        </dependency>
+
```
Comment (suhothayan) [inline component/pom.xml]: remove space

File: component/pom.xml [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.siddhi</groupId>
+            <artifactId>siddhi-extension-text-input-mapper</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.siddhi</groupId>
+            <artifactId>siddhi-extension-text-output-mapper</artifactId>
+            <version>4.0.0-SNAPSHOT</version>
```
Comment (suhothayan) [inline component/pom.xml]: No versions other than in parent pom

File: component/pom.xml [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.ops4j.pax.tipi</groupId>
+            <artifactId>org.ops4j.pax.tipi.hamcrest.core</artifactId>
+            <version>1.3.0.1</version>
```
Comment (suhothayan) [inline component/pom.xml]: No versions other than in parent pom

File: component/src/main/java/org/wso2/siddhi/extension/http/sink/util/HttpPayloadDataSource.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+    @Override
+    public void serializeData() {
+        try {
+            this.outputStream.write(this.value.getBytes(Charset.defaultCharset()));
+        } catch (IOException e) {
+            throw new HttpSinkAdaptorRuntimeException("Error occurred during writing the string message to the " +
+                    "output stream", e);
+        }
+
```
Comment (suhothayan) [inline component/src/main/java/org/wso2/siddhi/extension/http/sink/util/HttpPayloadDataSource.java]: remove unnecessary spaces

File: component/src/main/java/org/wso2/siddhi/extension/http/sink/util/HttpSinkUtil.java [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+     * @param mapType     user define map type.
+     * @param headersList list of headers defines by user.
+     * @return return the exact map type.
+     */
+    public String getContentType(String mapType, List<Header> headersList) {
+        for (Header h : headersList) {
+            if (HttpConstants.HTTP_CONTENT_TYPE.equals(h.getName())) {
+                return h.getValue();
+            }
+
```
Comment (suhothayan) [inline component/src/main/java/org/wso2/siddhi/extension/http/sink/util/HttpSinkUtil.java]: Remove space

File: component/src/main/java/org/wso2/siddhi/extension/http/sink/util/HttpSinkUtil.java [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+            SiddhiHttpSenderConfiguration httpsSender = new SiddhiHttpSenderConfiguration(clientStoreFile,
+                    clientStorePass);
+            httpsSender.setId(httpStaticProperties.get(HttpConstants.TO));
+            httpsSender.setScheme(httpStaticProperties.get(HttpConstants.PROTOCOL));
+            senderConf = new HashSet<>(Collections.singletonList(httpsSender));
+        } else {
+            SenderConfiguration httpSender = new SenderConfiguration(httpStaticProperties.get(HttpConstants.TO));
+            httpSender.setScheme(httpStaticProperties.get(HttpConstants.PROTOCOL));
+            senderConf = new HashSet<>(Collections.singletonList(httpSender));
+
```
Comment (suhothayan) [inline component/src/main/java/org/wso2/siddhi/extension/http/sink/util/HttpSinkUtil.java]: remove space

File: component/src/main/java/org/wso2/siddhi/extension/http/source/HttpSource.java [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+
+    @Override
+    public void resume() {
+        HttpConnectorRegistry.getInstance().getMessageProcessor(listenerUrl.replace(context, HttpConstants
+                .EMPTY_STRING)).resume();
+    }
+
+    @Override
+    public Map<String, Object> currentState() {
+
```
Comment (suhothayan) [inline component/src/main/java/org/wso2/siddhi/extension/http/source/HttpSource.java]: remove space

File: component/src/main/java/org/wso2/siddhi/extension/http/source/HttpWorkerThread.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+import org.wso2.carbon.messaging.CarbonCallback;
+import org.wso2.carbon.messaging.CarbonMessage;
+import org.wso2.siddhi.core.stream.input.source.SourceEventListener;
+import org.wso2.siddhi.extension.http.util.HttpConstants;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/** Handles the send data to source listener.
+ *
```
Comment (suhothayan) [inline component/src/main/java/org/wso2/siddhi/extension/http/source/HttpWorkerThread.java]: Format

File: component/src/main/java/org/wso2/siddhi/extension/http/source/util/HttpPausableThreadPoolExecutor.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+/**
+ * This is a custom implementation of the Thread pool executor to allow pausing and resuming of the thread pool.
+ */
+public class HttpPausableThreadPoolExecutor extends ThreadPoolExecutor {
+    private boolean paused;
+    private ReentrantLock lock;
+    private Condition condition;
+
+    public HttpPausableThreadPoolExecutor(int corePoolSize, int maximumPoolSize, long keepAliveTime, TimeUnit unit,
```
Comment (suhothayan) [inline component/src/main/java/org/wso2/siddhi/extension/http/source/util/HttpPausableThreadPoolExecutor.java]: Can't we just lock (talk to tishan)

File: component/src/main/java/org/wso2/siddhi/extension/http/source/util/HttpSourceUtil.java [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Set;
+
+
+/**
+ *
```
Comment (suhothayan) [inline component/src/main/java/org/wso2/siddhi/extension/http/source/util/HttpSourceUtil.java]: No comment !

File: component/src/main/java/org/wso2/siddhi/extension/http/source/util/HttpSourceUtil.java [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+        TransportProperty var3 = new TransportProperty();
+        var3.setName(HttpConstants.CLIENT_BOOTSTRAP_SOCKET_TIMEOUT);
+        var3.setValue(Integer.valueOf(sourceConfigReader.readConfig(HttpConstants.CLIENT_BOOTSTRAP_SOCKET_TIMEOUT,
+                HttpConstants.CLIENT_BOOTSTRAP_SOCKET_TIMEOUT_VALUE)));
+        properties.add(var3);
+        return new HashSet<>(properties);
+    }
+
+    /**
+     *
```
Comment (suhothayan) [inline component/src/main/java/org/wso2/siddhi/extension/http/source/util/HttpSourceUtil.java]: No comment!

File: pom.xml [hunk 13]
```diff
@@ -0,0 +1,457 @@
...
+            </dependency>
+            <dependency>
+                <groupId>org.wso2.carbon.caching</groupId>
+                <artifactId>org.wso2.carbon.caching</artifactId>
+                <version>${org.wso2.carbon.caching.version}</version>
+            </dependency>
+            <dependency>
+                <groupId>org.wso2.eclipse.osgi</groupId>
+                <artifactId>org.eclipse.osgi.services</artifactId>
+                <version>3.5.100.v20160504-1419</version>
```
Comment (suhothayan) [inline pom.xml]: move versions out

File: component/src/main/java/org/wso2/siddhi/extension/http/sink/HttpSink.java [hunk 14]
```diff
@@ -0,0 +1,305 @@
+/*
```
Comment (suhothayan) [inline component/src/main/java/org/wso2/siddhi/extension/http/sink/HttpSink.java]: packages need to be ../src/main/java/org/wso2/extension/siddhi/io/http/ ...

File: component/pom.xml (hunk 1 above)
Comment (Yasara123) [inline component/pom.xml]: Fixed in 0b31b37

File: component/pom.xml (hunk 2 above)
Comment (Yasara123) [inline component/pom.xml]: Fixed in  0b31b37

File: component/pom.xml (hunk 3 above)
Comment (Yasara123) [inline component/pom.xml]: Fixed in  0b31b37

File: component/pom.xml (hunk 4 above)
Comment (Yasara123) [inline component/pom.xml]: Fixed in  020c77c

File: component/src/main/java/org/wso2/siddhi/extension/http/sink/util/HttpPayloadDataSource.java (hunk 5 above)
Comment (Yasara123) [inline component/src/main/java/org/wso2/siddhi/extension/http/sink/util/HttpPayloadDataSource.java]: Fixed in 0b31b37

File: component/src/main/java/org/wso2/siddhi/extension/http/sink/util/HttpSinkUtil.java (hunk 7 above)
Comment (Yasara123) [inline component/src/main/java/org/wso2/siddhi/extension/http/sink/util/HttpSinkUtil.java]: Fixed in 0b31b37

File: component/src/main/java/org/wso2/siddhi/extension/http/sink/util/HttpSinkUtil.java (hunk 6 above)
Comment (Yasara123) [inline component/src/main/java/org/wso2/siddhi/extension/http/sink/util/HttpSinkUtil.java]: Fixed in 0b31b37

File: component/src/main/java/org/wso2/siddhi/extension/http/source/HttpSource.java (hunk 8 above)
Comment (Yasara123) [inline component/src/main/java/org/wso2/siddhi/extension/http/source/HttpSource.java]: Fixed in 0b31b37

File: component/src/main/java/org/wso2/siddhi/extension/http/source/HttpWorkerThread.java (hunk 9 above)
Comment (Yasara123) [inline component/src/main/java/org/wso2/siddhi/extension/http/source/HttpWorkerThread.java]: Fixed in 0b31b37

File: component/src/main/java/org/wso2/siddhi/extension/http/source/util/HttpSourceUtil.java (hunk 11 above)
Comment (Yasara123) [inline component/src/main/java/org/wso2/siddhi/extension/http/source/util/HttpSourceUtil.java]: Fixed in 0b31b37

File: component/src/main/java/org/wso2/siddhi/extension/http/source/util/HttpSourceUtil.java (hunk 12 above)
Comment (Yasara123) [inline component/src/main/java/org/wso2/siddhi/extension/http/source/util/HttpSourceUtil.java]: Fixed in 0b31b37

File: pom.xml (hunk 13 above)
Comment (Yasara123) [inline pom.xml]: Fixed in 0b31b37

File: component/src/main/java/org/wso2/siddhi/extension/http/sink/HttpSink.java (hunk 14 above)
Comment (Yasara123) [inline component/src/main/java/org/wso2/siddhi/extension/http/sink/HttpSink.java]: Fixed in 0b31b37

File: component/src/main/java/org/wso2/extension/siddhi/io/http/sink/util/HttpSinkUtil.java [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * {@code HttpSinkUtil } responsible of the all configuration reading and input formatting of http transport.
+ */
+public class HttpSinkUtil {
+    private static final Logger log = Logger.getLogger(HttpSinkUtil.class);
+
+    public HttpSinkUtil() {
```
Comment (tishan89) [inline component/src/main/java/org/wso2/extension/siddhi/io/http/sink/util/HttpSinkUtil.java]: What is the usage of his empty public constructor?

File: component/src/main/java/org/wso2/extension/siddhi/io/http/sink/util/HttpSinkUtil.java [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+        return senderConf;
+    }
+
+    /**
+     * Method is responsible for set transportation configuration values.
+     *
+     * @return return the set of netty transportation configuration.
+     */
+    public Set<TransportProperty> getTransportConfigurations(ConfigReader sourceConfigReader) {
+        ArrayList<TransportProperty> properties = new ArrayList<>();
```
Comment (tishan89) [inline component/src/main/java/org/wso2/extension/siddhi/io/http/sink/util/HttpSinkUtil.java]: Declaration should be a list

File: component/src/main/java/org/wso2/extension/siddhi/io/http/source/util/HttpSourceUtil.java [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+    public HttpSourceUtil() {
+    }
+
+    /**
+     * Method is responsible for set transportation configuration values.
+     *
+     * @return return the set of netty transportation configuration.
+     */
+    public Set<TransportProperty> getTransportConfigurations(ConfigReader sourceConfigReader) {
+        ArrayList<TransportProperty> properties = new ArrayList<>();
```
Comment (tishan89) [inline component/src/main/java/org/wso2/extension/siddhi/io/http/source/util/HttpSourceUtil.java]: Declaration should be a list

File: component/src/test/java/org/wso2/extension/siddhi/io/http/source/util/HttpTestUtil.java [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+
+/**
+ * Util class for test cases.
+ */
+public class HttpTestUtil {
+    private static final org.apache.log4j.Logger logger = org.apache.log4j.Logger.getLogger(HttpTestUtil.class);
+
+    public HttpTestUtil() {
```
Comment (tishan89) [inline component/src/test/java/org/wso2/extension/siddhi/io/http/source/util/HttpTestUtil.java]: What is the usage of pubic empty constructor

File: component/src/test/resources/activemq.properties [hunk 19]
```diff
@@ -0,0 +1,26 @@
+#
+# Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (tishan89) [inline component/src/test/resources/activemq.properties]: Wrong licence year

File: component/src/test/resources/conf/security/carbon-jaas.config [hunk 20]
```diff
@@ -0,0 +1,19 @@
+/*******************************************************************************
+*   Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (tishan89) [inline component/src/test/resources/conf/security/carbon-jaas.config]: Wrong licence year

File: component/src/test/resources/conf/security/permissions.yml [hunk 21]
```diff
@@ -0,0 +1,141 @@
+################################################################################
+#   Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (tishan89) [inline component/src/test/resources/conf/security/permissions.yml]: Wrong licence year

Comment (tishan89) [reply component/src/test/resources/conf/security/permissions.yml]: Fix for all below occurrences.

File: tests/osgi-tests/src/test/java/org/wso2/extension/siddhi/io/http/test/osgi/BasicAuthTrue.java [hunk 22]
```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (tishan89) [inline tests/osgi-tests/src/test/java/org/wso2/extension/siddhi/io/http/test/osgi/BasicAuthTrue.java]: Fix licence year

File: tests/osgi-tests/src/test/java/org/wso2/extension/siddhi/io/http/test/osgi/server/HTTPServer.java [hunk 23]
```diff
@@ -0,0 +1,128 @@
+package org.wso2.extension.siddhi.io.http.test.osgi.server;
```
Comment (tishan89) [inline tests/osgi-tests/src/test/java/org/wso2/extension/siddhi/io/http/test/osgi/server/HTTPServer.java]: Package should come after licence and licence year should be fixed

Comment (tishan89) [reply tests/osgi-tests/src/test/java/org/wso2/extension/siddhi/io/http/test/osgi/server/HTTPServer.java]: Same for below classes

File: tests/osgi-tests/src/test/java/org/wso2/extension/siddhi/io/http/test/osgi/util/ServerUtil.java [hunk 24]
```diff
@@ -0,0 +1,62 @@
+package org.wso2.extension.siddhi.io.http.test.osgi.util;
```
Comment (tishan89) [inline tests/osgi-tests/src/test/java/org/wso2/extension/siddhi/io/http/test/osgi/util/ServerUtil.java]: Add licence

Comment (tishan89) [reply tests/osgi-tests/src/test/java/org/wso2/extension/siddhi/io/http/test/osgi/util/ServerUtil.java]: Same applies to classes below

File: component/src/main/java/org/wso2/extension/siddhi/io/http/sink/util/HttpSinkUtil.java (hunk 15 above)
Comment (Yasara123) [reply component/src/main/java/org/wso2/extension/siddhi/io/http/sink/util/HttpSinkUtil.java]: Fixed in 156b93ab839dfc225e0a915f8951b61db7afa2ab

File: component/src/main/java/org/wso2/extension/siddhi/io/http/sink/util/HttpSinkUtil.java (hunk 16 above)
Comment (Yasara123) [reply component/src/main/java/org/wso2/extension/siddhi/io/http/sink/util/HttpSinkUtil.java]: Fixed in 156b93ab839dfc225e0a915f8951b61db7afa2ab

File: component/src/main/java/org/wso2/extension/siddhi/io/http/source/util/HttpSourceUtil.java (hunk 17 above)
Comment (Yasara123) [reply component/src/main/java/org/wso2/extension/siddhi/io/http/source/util/HttpSourceUtil.java]: Fixed in 156b93ab839dfc225e0a915f8951b61db7afa2ab

File: tests/osgi-tests/src/test/java/org/wso2/extension/siddhi/io/http/test/osgi/server/HTTPServer.java (hunk 23 above)
Comment (Yasara123) [reply tests/osgi-tests/src/test/java/org/wso2/extension/siddhi/io/http/test/osgi/server/HTTPServer.java]: Fixed in 0b31b3775a3e0156cd5309071629ff19e455bc29

File: component/src/test/java/org/wso2/extension/siddhi/io/http/source/util/HttpTestUtil.java (hunk 18 above)
Comment (Yasara123) [reply component/src/test/java/org/wso2/extension/siddhi/io/http/source/util/HttpTestUtil.java]: Fixed in 6bf095af9d7a4c00220069badb50517cec553139

File: component/src/test/resources/activemq.properties (hunk 19 above)
Comment (Yasara123) [reply component/src/test/resources/activemq.properties]: Fixed in 6bf095af9d7a4c00220069badb50517cec553139

File: component/src/test/resources/conf/security/carbon-jaas.config (hunk 20 above)
Comment (Yasara123) [reply component/src/test/resources/conf/security/carbon-jaas.config]: Fixed in 6bf095af9d7a4c00220069badb50517cec553139

File: component/src/test/resources/conf/security/permissions.yml (hunk 21 above)
Comment (Yasara123) [reply component/src/test/resources/conf/security/permissions.yml]: Fixed in 6bf095af9d7a4c00220069badb50517cec553139

File: tests/osgi-tests/src/test/java/org/wso2/extension/siddhi/io/http/test/osgi/BasicAuthTrue.java (hunk 22 above)
Comment (Yasara123) [reply tests/osgi-tests/src/test/java/org/wso2/extension/siddhi/io/http/test/osgi/BasicAuthTrue.java]: Fixed in 6bf095af9d7a4c00220069badb50517cec553139

File: tests/osgi-tests/src/test/java/org/wso2/extension/siddhi/io/http/test/osgi/util/ServerUtil.java (hunk 24 above)
Comment (Yasara123) [reply tests/osgi-tests/src/test/java/org/wso2/extension/siddhi/io/http/test/osgi/util/ServerUtil.java]: Fixed in 6bf095af9d7a4c00220069badb50517cec553139

**Pull request merged**